int32 state
---
int32 action
bool success
string error_msg